### DISCUSSION TITLE
How to use Taxonomies/Terms in Forms

### DISCUSSION THREAD

**Discussion by pathly-dev** [statamic/cms#6510] [Q&A] [answered]
Hey, I want to create a form with a select where the frontend user can choose between options that come from my taxonomy.
I have something similar on the panel, where I used the terms fieldtype, so the panel user can choose a option of the given taxonomy. But it seems like thats not working the same way.
This is what I have so far for my form:
```
{{ form:story-upload }}
  {{ fields }}
    <div>
      <label>{{ instructions }}</label>
      {{ if type == "terms" }}
        <select>
          {{ taxonomies }}
            <option value="">{{ title }}</option>
          {{ /taxonomies }}
        </select>
      {{ else }}
        <div>{{ field }}</div>
      {{ /if }}
    </div>
  {{ /fields }}
{{ /form:story-upload }}
```
But no taxonomy options are shown and also the {{ title }} is not the title of the taxonomy but the forms title.

**Comment by byron-roots** [comment, marked as answer]
You can make the Taxonomy Terms field type available to use by adding `\Statamic\Fieldtypes\Terms::makeSelectableInForms();` your AppServiceProvider boot method.

Then you just need to create a view in `resources/views/vendor/statamic/forms/fields/term.antlers.html` and then you can do something like
``` antlers
<div class="o-form__select">
    <label class="u-sr" for="{{ handle }}">
        {{ display }}
    </label>
    <select id="{{ handle }}" class="c-select c-select--md"  name="{{ handle }}[]" {{ (validate | contains:required) ? 'required="required"' : '' }} {{ multiple ?= "multiple" }}>
        <option value>{{ if instructions }}{{ instructions }}{{ else }}Please select...{{ /if }}</option>
        {{ taxonomy from="{collections|option_list}" }}
            <option value="{{ id }}" {{ if old|in_array:id || old === id }}selected{{ /if }}>{{ title }}</option>
        {{ /taxonomy }}
    </select>
    {{ if error }}
        <span class="c-status c-status--error">
            {{ error }}
        </span>
    {{ /if }}
</div>
```

I've only done this with collections but that above should work, might need a few tweaks though

**Reply by pathly-dev** [reply]
> You can make the Taxonomy Terms field type available to use by adding `\Statamic\Fieldtypes\Terms::makeSelectableInForms();` your AppServiceProvider boot method.
> 
> Then you just need to create a view in `resources/views/vendor/statamic/forms/fields/term.antlers.html` and then you can do something like
> 
> ```
> <div class="o-form__select">
>     <label class="u-sr" for="{{ handle }}">
>         {{ display }}
>     </label>
>     <select id="{{ handle }}" class="c-select c-select--md"  name="{{ handle }}[]" {{ (validate | contains:required) ? 'required="required"' : '' }} {{ multiple ?= "multiple" }}>
>         <option value>{{ if instructions }}{{ instructions }}{{ else }}Please select...{{ /if }}</option>
>         {{ taxonomy from="{collections|option_list}" }}
>             <option value="{{ id }}" {{ if old|in_array:id || old === id }}selected{{ /if }}>{{ title }}</option>
>         {{ /taxonomy }}
>     </select>
>     {{ if error }}
>         <span class="c-status c-status--error">
>             {{ error }}
>         </span>
>     {{ /if }}
> </div>
> ```
> 
> I've only done this with collections but that above should work, might need a few tweaks though

This also doesnt seem to work. I get the error "Taxonomy [] not found". 
I think this `{{ taxonomy from="{collections|option_list}" }}` is the problem, but this is what I actually was asking for. What exactly do I need to write here? I already tried different versions of `{{ taxonomy }}`, `{{ taxonomies  }}`, `{{ taxonomy:{taxonomies} }}`. I dont understand how to get the taxonomies that are specified for the specific forms field.

**Reply by byron-roots** [reply]
Ahh forgot to update that line. In the field type you can choose which Taxonomies it uses, so you'll need to select at least one. Then I believe the line should read `{{ taxonomy from="{taxonomies|option_list}" }}`. The `{taxonomies|option_list}` will output a comma separated list for you, so if you choose more than one it will still work

**Reply by pathly-dev** [reply]
Ahh yes that's working, thanks a lot! 
I'm just a little bit confused why sometimes you need the version with the from= and sometimes the version with the colon.

**Reply by byron-roots** [reply]
The colon format is just shorthand to make things tidy.`{{ taxonomy:tags }}` is the same as `{{ taxonomy from="tags" }}`. As we are using a dynamic input and a modifier using the long hand format is more readable, and I'm not sure the shorthand would work with the dynamic input and modifier

**Reply by pathly-dev** [reply]
Okay I understand, thanks!

**Reply by pathly-dev** [reply]
@byron-roots I have another question. This all works fine on the frontend site. But when I submit the form, I do receive all informations I typed in in input fields but I don't receive the information of the selected option of a select.

**Reply by byron-roots** [reply]
Could you post the view file you have for the field?

**Reply by pathly-dev** [reply]
```
<section class="section_form">
  {{ form:story-upload }}
    {{ fields }}
      {{ if type == "terms" }}
        <select>
          <option disabled selected>Auswählen</option>
          {{ taxonomy from="{taxonomies|option_list}" }}
            <option value="{{ slug }}" name="{{ slug }}">{{ title }}</option>
          {{ /taxonomy }}
        </select>
      {{ else }}
        <div class="{{ type }}-input">
          {{ if instructions }}<label>{{ instructions }}</label>{{ /if }}
          {{ field }}
        </div>
      {{ /if }}
    {{ /fields }}

    <button type="submit">Submit</button>
  {{ /form:story-upload }}
</section>
```

**Reply by byron-roots** [reply]
You'll want to pass the ID as the option value, and you'll need to set a name field for select field

``` antlers
<select name="{{ handle }}[]" {{ multiple ?= "multiple" }}>
    {{ taxonomy from="{taxonomies|option_list}" }} 
        <option value="{{ id }}" name="{{ slug }}">{{ title }}</option>
    {{ /taxonomy }}
</select>
```

**Reply by pathly-dev** [reply]
Ah ok, thanks thats working! 
And how do you do that with the "multiple"? Where does the information whether it should be multiple or not come from? There is no option on the terms fieldtype.

**Reply by byron-roots** [reply]
Ah yeah there isn't, think that was just left over from my usual select field. You could probably do a check for `max_items` instead.  Something like `{{ max_items > 1 ?= 'multiple' }} `

**Reply by pathly-dev** [reply]
Thanks a lot!

**Reply by pathly-dev** [reply]
@byron-roots Sorry for bothering you with all this questions. But I have another question.
Since just looping through alle the fields isn't flexibel enough for my project, I thought about to do it manually like this
```
{{ form:myform }}
  <div class="text-input">
    <label>Vorname</label>
    <input type="text" name="firstname" value="{{ old:firstname }}">
  </div>
{{ /form:myform }}
```
But I'm not sure how to do it in combination with the select and my predefined options that I defined in the blueprint of the form. How do I call the field and loop through the options?
I already tried things like this
```
{{ handle-of-my-field }}
    {{ field }}
{{ /handle-of-my-field }}
```
and this
```
{{ handle-of-my-field }}
  <select>
    <option value="{{ options }}">
  </select>
{{ /handle-of-my-field }}
```
and this
```
<select name="handle-of-my-field">
  {{ taxonomy from="{taxonomies|option_list}" }}
      <option value="{{ id }}" name="{{ slug }}">{{ title }}</option>
  {{ /taxonomy }}
</select>
```
But I haven't found the correct way to write it. Unfortunately the manual writing version isn't really good described in the Docs

**Reply by byron-roots** [reply]
You'd still need to use the field loop to access the options

``` antlers
{{ fields }}
    {{ if handle == handle-of-my-field }}
        <select name="handle-of-my-field">
            {{ taxonomy from="{taxonomies|option_list}" }}
            <option value="{{ id }}" name="{{ slug }}">{{ title }}</option>
            {{ /taxonomy }}
        </select>
    {{ /if }}
{{ /fields }}
```

It may be better to set the taxonomies option in your collection blueprint, instead of the form blueprint. You can leave the terms field in form blueprint open to all taxonomies as this will be a read only file. It would then be easier to fetch this in your template. 

Not sure what limitations were causing issues though so it may not be useful for yu.

**Reply by pathly-dev** [reply]
So the project I'm working on is to implement this form where users can fill in their stories and after submitting the submission is supposed to be a draft in my collection "stories" (How this is going to work is a different thing :D). So stories can be either created through the "create entry" in the collection or via the form submission. So I thought that I need exactly all the fields of my stories blueprint in my form blueprint in the exact same way. So I have e.g. two different terms fields in my stories blueprint that are defined with two different taxonomies and I also have radio buttons where the options are predefined in my stories blueprint, so i thought I need to predefine them exactly the same way in the form blueprint. But if I understand correctly, I also can just let the taxonomy definition open and just loop through the taxonomies in general? So would it work the same with radio buttons? Dont I have to predefined the options in the form blueprint? Would it also be possible to just list the options like 
```
<input type="radio" name="here-my-handle-of-the-form-field?" value="option1">option 1</option>
<input type="radio" name="here-my-handle-of-the-form-field?" value="option2">option 2</option>
```

**Reply by byron-roots** [reply]
The options are there to limit what you can do when adding entries. As the form fields are read only it doesn't really matter, unless you want to use the options in the form template. If you already know what terms/taxonomies you want to use on then it's not really necessary on the form terms field.

``` antlers
<select name="handle-of-my-field">
    {{ taxonomy:categories }}
       <option value="{{ id }}" name="{{ slug }}">{{ title }}</option>
    {{ /taxonomy:categories }}
</select>
```

Hopefully that makes sense

**Reply by pathly-dev** [reply]
Ok got it, thanks! :)

**Reply by pathly-dev** [reply]
So actually the type of the fields in the form blueprint could be text even if it's in the frontend actually a select/terms, right? Because in the end the field just gets the value of the selected item. Or should this still be the type that it's supposed to be?

**Reply by byron-roots** [reply]
It could be a text field, but as you want to save the ID of the term (as this is how it would work in your entries) the terms field will display the actual term in Form submission. If you don't ever need to see the form submissions a text field would be fine.

If you aren't going to look at the form submissions, and are only using the form to send data, then you could probably just create a custom controller and route to handle the form submission and creating the entry

**Reply by pathly-dev** [reply]
Okay yes makes sense. Thanks for all the answers :)
